Cpp_test/lib/ipc-shm-send.cpp

```diff
@@ -0,0 +1,159 @@
+#include "ipc-shm-send.h"
+#include <iostream>
+#include <cassert>
+
+IPCShmSend::~IPCShmSend()
+{
+
+	if (this->shm_ptr != nullptr)
+	{
+		if (this->shm_ptr->checkLockMutex == true)
+		{
+			this->shm_ptr->checkLockMutex = false;
+		}
+		int ret1 = pthread_cond_destroy(&(this->shm_ptr->cond_re));
+		if (ret1 != 0)
+		{
+			std::cerr << "Error at pthread_cond_destroy(): receive " << strerror(ret1) << std::endl;
+		}
+		sleep(3);
+		
+		int ret2 = pthread_mutex_trylock(&(this->shm_ptr->mutex));
+		if (ret2 != 0)
+		{
+			std::cerr << "Error at pthread_mutex_trylock(): " << strerror(ret2) << std::endl;
+		}
+		else
+		{
+			int ret3 = pthread_mutex_unlock(&(this->shm_ptr->mutex));
+			if (ret3 != 0)
+			{
+				std::cerr << "Error at pthread_mutex_unlock(): " << strerror(ret3) << std::endl;
+			}
+			else
+			{
+				int ret4 = pthread_mutex_destroy(&(this->shm_ptr->mutex));
+				if (ret4 != 0)
+				{
+					std::cerr << "Error at pthread_mutex_destroy(): " << strerror(ret4) << std::endl;
+				}
+			}
+		}
+
+		int ret5 = pthread_mutexattr_destroy(&(this->mutex_attr));
+		if (ret5 != 0)
+		{
+			std::cerr << "Error at pthread_mutexattr_destroy(): " << strerror(ret5) << std::endl;
+		}
+	}
+}
+
+void IPCShmSend::init()
+{
+	this->file_handler.open_file();
+	this->open_shm();
+	this->set_shm_size();
+	this->map_shm();
+	if (this->shmd > 0)
+	{
+		close(this->shmd);
+	}
+	pthread_mutexattr_destroy(&(this->mutex_attr));
+	this->init_mutex();
+
+	pthread_condattr_destroy(&(this->cond_attr));
+	this->init_cond();
+}
+
+void IPCShmSend::transfer()
+{
+	long read_bytes{0};
+	unsigned long long file_size{this->file_handler.get_file_size()};
+	if (file_size == 0)
+	{
+		throw std::runtime_error("ERROR: File size = 0.");
+	}
+
+	unsigned long long total_sent_bytes{0};
+	std::cout << "Waiting for receiver..." << std::endl;
+	while (!this->shm_ptr->is_end)
+	{
+		if (this->shm_ptr->checkLockMutex == true)
+		{
+			lock_mutex();
+		}
+
+		while (this->shm_ptr->data_version_received != this->shm_ptr->data_version)
+		{
+
+			if (pthread_cond_wait(&(this->shm_ptr->cond_re), &(this->shm_ptr->mutex)) != 0)
```

Comment:
Maybe you should also check there the status of `checkLockMutex` ?

Reply:
Here, if it is true, will lock the mutex, else it will throw with error pthread_cond_wait(). Is it ok?

Reply:
I think if it's true you can safely do pthread_cond_wait (which involves a lock), but if it's false you should throw yes. Maybe it is what you are saying.

Reply:
(both comments are valid for the receiver too)

Reply:
Yes, it is what I mean

Reply:
I got a segmentation fault here, I think because it already lock, now checkLockMutex is still true and setting lock again is impossible.
![image](https://user-images.githubusercontent.com/111491602/208914445-9eedb54a-a4dc-4378-b042-79b35fde3f5e.png)

Reply:
You don't have to lock it again, it's just to protect the pthread_cond_wait() function
```c++
while (!this->shm_ptr->is_end)
	{
		if (this->shm_ptr->checkLockMutex == true)
		{
			lock_mutex();
		} else 
		{
			throw ...
		}


		while (this->shm_ptr->data_version_received != this->shm_ptr->data_version)
		{
			if (this->shm_ptr->checkLockMutex == false)
			{
				throw ...
			}
			if (pthread_cond_wait(&(this->shm_ptr->cond_re), &(this->shm_ptr->mutex)) != 0)
			{
				throw std::runtime_error("ERROR: pthread_cond_wait() receive.");
			}
		}
```